File: frontend/budget-managment/src/types/projet.ts
export type ProjetStatus = "brouillon" | "soumis" | "approuve" | "rejete" | "en_execution" | "cloture";

export interface Projet {
  id: number;
  code: string;
  titre: string;
  description?: string | null;
  objectif?: string | null;
  resultat_attendu?: string | null;
  statut: ProjetStatus;
  date_creation?: string;
  date_modification?: string;
  date_debut_prevue?: string | null;
  date_fin_prevue?: string | null;
  cout_estime: number;
  budget_realise_total?: number;
  departement_id: number;
  exercice_id: number;
  created_by_id: number;
  chef_projet_id?: number | null;
  departement?: {
    id: number;
    nom: string;
  };
  chef_projet?: {
    id: number;
    nom: string;
    prenom?: string | null;
    email: string;
  };
  created_by?: {
    id: number;
    nom: string;
    prenom?: string | null;
    email: string;
  };
}

export interface ProjetCreate {
  code: string;
  titre: string;
  description?: string;
  objectif?: string;
  resultat_attendu?: string;
  date_debut_prevue?: string;
  date_fin_prevue?: string;
  cout_estime?: number;
  departement_id: number;
  exercice_id: number;
  created_by_id: number;
  chef_projet_id: number;
}

export interface ProjetUpdate {
  code?: string;
  titre?: string;
  description?: string;
  objectif?: string;
  resultat_attendu?: string;
  statut?: ProjetStatus;
  date_debut_prevue?: string;
  date_fin_prevue?: string;
  cout_estime?: number;
  budget_realise_total?: number;
  departement_id?: number;
  exercice_id?: number;
  created_by_id?: number;
  chef_projet_id?: number;
}

File: backend/app/models/projet.py
from datetime import datetime

from sqlalchemy import Column, Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import relationship

from app.database import Base


class Projet(Base):
    __tablename__ = "projets"

    id = Column(Integer, primary_key=True, index=True)
    code = Column(String(100), unique=True, index=True, nullable=False)
    titre = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    objectif = Column(Text, nullable=True)
    statut = Column(String(50), default="brouillon")
    date_creation = Column(DateTime, default=datetime.utcnow)
    date_modification = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    date_debut_prevue = Column(Date, nullable=True)
    date_fin_prevue = Column(Date, nullable=True)
    resultat_attendu = Column(Text, nullable=True)
    cout_estime = Column(Numeric(15, 2), default=0)
    budget_realise_total = Column(Numeric(15, 2), default=0)

    departement_id = Column(Integer, ForeignKey("departements.id"), nullable=False)
    exercice_id = Column(Integer, ForeignKey("exercices_budgetaires.id"), nullable=False)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    chef_projet_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    departement = relationship("Departement", back_populates="projets")
    exercice = relationship("ExerciceBudgetaire", back_populates="projets")
    created_by = relationship("User", back_populates="projets_crees", foreign_keys=[created_by_id])
    chef_projet = relationship("User", back_populates="projets_geres", foreign_keys=[chef_projet_id])
    budget = relationship("Budget", back_populates="projet", uselist=False)
    mouvements_financiers = relationship(
        "MouvementFinancier",
        back_populates="projet",
        cascade="all, delete-orphan",
    )
